numba/interpreter.py

```diff
@@ -168,59 +213,54 @@ def interpret(self):
+        # post-processing
+        self._post_processing()
+
+    @utils.cached_property
+    def variable_lifetime(self):
+        return VariableLifetime(self.blocks)
+
+    def _post_processing(self):
         """
-        Compute the live variables at the beginning of each block
-        and at each yield point.
-        The ``var_def_map`` and ``var_dead_map`` indicates the variable defined
-        and deleted at each block, respectively.
+        Post-processing and analysis on generated IR
         """
-        # live var at the entry per block
-        block_entry_vars = collections.defaultdict(set)
-
-        def fix_point_progress():
-            return tuple(map(len, block_entry_vars.values()))
-
-        cfg = self.cfa.graph
-        old_point = None
-        new_point = fix_point_progress()
-
-        # Propagate defined variables and still live the successors.
-        # (note the entry block automatically gets an empty set)
-
-        # Note: This is finding the actual available variables at the entry
-        #       of each block. The algorithm in _compute_live_map() is finding
-        #       the variable that must be available at the entry of each block.
-        #       This is top-down in the dataflow.  The other one is bottom-up.
-        while old_point != new_point:
-            # We iterate until the result stabilizes.  This is necessary
-            # because of loops in the graphself.
-            for offset in self.blocks:
-                # vars available + variable defined
-                avail = block_entry_vars[offset] | var_def_map[offset]
-                # substract variables deleted
-                avail -= var_dead_map[offset]
-                # add ``avail`` to each successors
-                for succ, _data in cfg.successors(offset):
-                    block_entry_vars[succ] |= avail
-
-            old_point = new_point
-            new_point = fix_point_progress()
+        # emit del nodes
+        self._insert_var_dels()
 
+        vlt = self.variable_lifetime
+        bev = analysis.compute_live_variables(vlt.cfg, self.blocks,
+                                              vlt.usedefs.defmap,
+                                              vlt.deadmaps.combined)
         for offset, ir_block in self.blocks.items():
-            self.block_entry_vars[ir_block] = block_entry_vars[offset]
+            self.block_entry_vars[ir_block] = bev[offset]
+
+        if self.generator_info:
+            self._compute_generator_info()
+
+    def _populate_generator_info(self):
+        """
+        Fill `index` for the Yield instruction and create YieldPoints.
+        """
+        dct = self.generator_info.yield_points
+        assert not dct, 'rerunning _populate_generator_info'
+        for block in self.blocks.values():
+            for inst in block.body:
+                if isinstance(inst, ir.Assign):
+                    yieldinst = inst.value
+                    if isinstance(yieldinst, ir.Yield):
```

Comment:
It shouldn't be a problem.  Since the IR comes from the bytecode, the YIELD_VALUE bytecode is an operation that store a value on the stack regardless of its usage (subexpression or not).  The IR will always be in form of `$xx = yield ...` ; thus always a Yield in Assign.